Tls

xtreme-sameer-vohra opened the pull request:
Enable NATS TLS connections with CA verification

Ref Story - https://www.pivotaltracker.com/story/show/139228793

andrew-su committed:
Add TLS client

- tls client supports provided CA cert
- It currently swallows the INFO msg

[#139228793](https://www.pivotaltracker.com/story/show/139228793)

Signed-off-by: Sameer Vohra <svohra@pivotal.io>

xtreme-sameer-vohra committed:
Refactored ConnectionInfo to accept a CertPool

- Removed unnecessary TLSConnectionInfo struct.
- Updated tests.

[#139228793](https://www.pivotaltracker.com/story/show/139228793)

Signed-off-by: Andrew Su <asu@pivotal.io>

Commit:
Update README with instructions for TLS connection

[#139228793](https://www.pivotaltracker.com/story/show/139228793)

Signed-off-by: Sameer Vohra <svohra@pivotal.io>

dpb587-pivotal commented on connection.go:
```diff
@@ -62,15 +64,55 @@ func NewConnection(addr, user, pass string) *Connection {
+			return nil, err
+		}
+
+		// TODO Save Info packet somewhere accessible
```
is this still a todo? once tls is established, does server send another info packet, in which case we don't care about this one?

xtreme-sameer-vohra replied:
We'll remove the TODO for now because the non-tls connection is also swallowing the INFO msg atm.

dpb587-pivotal commented on connection_integration_test.go:
```diff
@@ -0,0 +1,120 @@
+	client := NewClient()
+
+	roots := x509.NewCertPool()
+    ok := roots.AppendCertsFromPEM([]byte(CA))
```
go fmt

dpb587-pivotal commented on connection.go:
```diff
@@ -62,15 +64,55 @@ func NewConnection(addr, user, pass string) *Connection {
+		br := bufio.NewReaderSize(conn, 32768)
+		_, err = Parse(br)
+		if err != nil {
+			return conn, err
```
probably shouldn't return `conn` (here and 84), just to avoid somebody accidentally ignoring an error and then blindly using an unexpectedly insecure connection

xtreme-sameer-vohra replied:
Fair, will address that

dpb587-pivotal commented on connection.go:
```diff
@@ -62,15 +64,55 @@ func NewConnection(addr, user, pass string) *Connection {
 func (c *ConnectionInfo) ProvideConnection() (*Connection, error) {
-	conn := NewConnection(c.Addr, c.Username, c.Password)
+	var conn *Connection
+	if c.CertPool == nil || len(c.CertPool.Subjects()) == 0 {
```
should we have the length check in this? since non-nil CertPool is non-default, it seems safer to say "if CertPool is present in any way... we should act securely". maybe

xtreme-sameer-vohra committed:
More defensive err handling in TLS Dial

- gofmt on connection & connection_integration

[#139228793](https://www.pivotaltracker.com/story/show/139228793)

Signed-off-by: Dale Wick <dwick@pivotal.io>

dpb587-pivotal commented on connection_integration_test.go:
```diff
@@ -109,12 +109,28 @@ func (t *TLSSuite) TestNewTLSConnectionWithWrongCA(c *C) {
+func (t *TLSSuite) TestNewTLSConnectionWithEmptyCertPool(c *C) {
+	client := NewClient()
+
+	roots := x509.NewCertPool()
```
could inline this since `roots` isn't used elsewhere

xtreme-sameer-vohra replied:
Done

dpb587-pivotal commented on connection_integration_test.go:
```diff
@@ -0,0 +1,136 @@
+
+}
+
+func (t *TLSSuite) TestNewTLSConnectionWithEmptyCertPool(c *C) {
```
👍

xtreme-sameer-vohra committed:
Cleaned up test function.

[#139228793](https://www.pivotaltracker.com/story/show/139228793)

Signed-off-by: Dale Wick <dwick@pivotal.io>

Commit:
Bump travis to go 1.7, 1.8

[#139228793](https://www.pivotaltracker.com/story/show/139228793)

Signed-off-by: Sameer Vohra <svohra@pivotal.io>
